File: internal/tests/resources/test_book.go
package resources_test

// This file migrates tests/resources/test_book.py, which verified that
// GET /books/1 returns HTTP 200 with the expected book JSON payload.
//
// The original test relied on a pytest `client` fixture (a Flask test client)
// and accessed the decoded body via `response.json`. In Go the idiomatic
// equivalent is to spin up the application handler with httptest, issue a
// real request, and decode the JSON response body explicitly.
//
// MIGRATION_NOTE: The Python test imported the resource module via a wildcard
// import (`from resources.book import *`) purely so that its route decorators
// registered the endpoints as a side effect. Go has no decorator-based route
// registration; routes are wired explicitly through
// resources.RegisterRoutes, which newTestHandler (internal/conftest.go)
// performs. We therefore call that shared helper instead of relying on any
// import side effect.

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	internal "bookapi/internal"
)

// TestGetBook verifies that GET /books/1 responds with HTTP 200 and the
// expected book payload {"id": 1, "title": "Python for Dummies"}.
func TestGetBook(t *testing.T) {
	handler := internal.NewTestHandler(t)

	srv := httptest.NewServer(handler)
	defer srv.Close()

	req, err := http.NewRequest(http.MethodGet, srv.URL+"/books/1", nil)
	if err != nil {
		t.Fatalf("building request: %v", err)
	}

	resp, err := srv.Client().Do(req)
	if err != nil {
		t.Fatalf("performing request: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		t.Fatalf("status code = %d, want %d", resp.StatusCode, http.StatusOK)
	}

	var got struct {
		ID    int    `json:"id"`
		Title string `json:"title"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&got); err != nil {
		t.Fatalf("decoding response body: %v", err)
	}

	want := struct {
		ID    int
		Title string
	}{
		ID:    1,
		Title: "Python for Dummies",
	}

	if got.ID != want.ID {
		t.Errorf("id = %d, want %d", got.ID, want.ID)
	}
	if got.Title != want.Title {
		t.Errorf("title = %q, want %q", got.Title, want.Title)
	}
}

File: src/tests/resources/test_book.ts
import request from 'supertest';
import { Express } from 'express';
import { getApp, createTestHarness, TestHarness } from '../../conftest';

/**
 * Integration tests for the Book resource.
 *
 * MIGRATION_NOTE: The source (`tests/resources/test_book.py`) used a pytest
 * `client` fixture (a Flask test client) and a wildcard import
 * (`from resources.book import *`) whose only purpose was the import side
 * effect of registering the book routes. In this project route registration
 * is done explicitly by the `Server`/app, so we obtain the Express app via
 * `getApp()` and drive it with supertest instead of a Flask test client.
 *
 * MIGRATION_NOTE: The Python `client` fixture is replaced by the shared
 * `{ client, repo }` harness (`createTestHarness`). Here only the HTTP client
 * side is needed, so the harness is built and the app is used with supertest.
 */
describe('GET /books/1', () => {
  let app: Express;
  let harness: TestHarness;

  beforeAll(async () => {
    harness = await createTestHarness();
    app = getApp();
  });

  afterAll(async () => {
    if (harness && typeof (harness as unknown as { close?: () => Promise<void> }).close === 'function') {
      await (harness as unknown as { close: () => Promise<void> }).close();
    }
  });

  it('returns 200 and the expected book payload', async () => {
    const response = await request(app).get('/books/1');

    expect(response.status).toBe(200);
    expect(response.body).toEqual({
      id: 1,
      title: 'Python for Dummies',
    });
  });
});
